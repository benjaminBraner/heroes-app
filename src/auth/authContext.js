import { createContext } from "react";

//el nombre esta capitalizado porque en pocas palabras es un componente, porque el createContext genera un componenente
export const  AuthContext = createContext();
