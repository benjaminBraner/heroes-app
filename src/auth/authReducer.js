import { types } from "../types/types";



/*es una simple funcion que recibe el state y un action que es el que va a modificar el state*/
export const authReducer = (state = {}, action) => {

     switch (action.type) {
          case types.login:

               return {
                    ...action.payload,
                    logged: true
               }
          
          case types.logout:
               
               return {
                    logged: false
               }
               
          default:
               return state;
     }


}