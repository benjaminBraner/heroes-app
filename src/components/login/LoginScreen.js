import { useContext } from "react";
import { useNavigate } from "react-router-dom";
import { AuthContext } from "../../auth/authContext";
import { useForm } from "../../hooks/useForm";
import { types } from "../../types/types";


export const LoginScreen = () => {

     /*en la v5 de react-router-dom se recibia en los props un history y otros elementos que me ayudaban al redireccionamiento
     entre pantallas, pero ahora en la v6 se hace mediante el useNavigate*/
     const navigate = useNavigate();
     const { dispatch } = useContext( AuthContext );

     const [ { inputName }, handleInputChange ] = useForm( { inputName: "" } );

     const handleLogin = () => {    
          if ( inputName ) {
               dispatch({
                    type: types.login,
                    payload: {
                         name: inputName
                    }
               })
               /* localStorage.getItem("lastPath") ese lastPath lo defiini en el <PrivateRoute/> si es que existe seria la ultima url 
               que el usuario visito antes de hacer el logout*/
               const path = localStorage.getItem( "lastPath" ) || "/marvel";
               /*navigate( path, { replace: true } );   dentro de los parentesis es la ruta a la que quiero redireccionar, la diferencia 
               es que esta en vez de redireccionar, remplaza la pagina actual, osea como si no hubiera visitado la pagina en la que estaba*/
               navigate( path, { replace: true } );
          }
          
          
     }

     return (
          <div className="container mt-5">
               <h1>Login Screen</h1>
               <hr />

                    <form>
                         <input
                              type="text"
                              name="inputName"
                              onChange={ handleInputChange }
                              value={ inputName }
                              autoComplete="off"
                         />

                         <button 
                              type="submit"
                              className="btn btn-primary"
                              onClick={ handleLogin }
                         >
                              Login
                         </button>

                    </form>

          </div>
     )
}
