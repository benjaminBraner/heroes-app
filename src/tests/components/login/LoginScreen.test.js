import { mount } from 'enzyme'
import { MemoryRouter } from 'react-router-dom'
import { AuthContext } from '../../../auth/authContext'
import { LoginScreen } from '../../../components/login/LoginScreen'
import { types } from '../../../types/types'

const mockNavigate = jest.fn();

jest.mock( 'react-router-dom' , () => ({
     ...jest.requireActual('react-router-dom'),
     useNavigate: () => mockNavigate
}))

describe('Pruebas en el LoginComponent', () => {

     

     const contextValue = {
          user: {
               logged: false
          },
          dispatch: jest.fn()
     }
     
     const wrapper = mount(
          <AuthContext.Provider value={ contextValue }>
               <MemoryRouter initialEntries={ ['/login'] }>
                    <LoginScreen />
               </MemoryRouter>
          </AuthContext.Provider>
     )

     test('debe mostrarse correctamente', () => {
          expect( wrapper ).toMatchSnapshot()
          
     })
     

     test('debe realizar el dispatch y la navegacion', () => {
          
          wrapper.find( 'input' ).simulate( 'change' , {
               target: {
                    name: 'inputName',
                    value: 'Pedro Parques'
               }
          } )

          const handleClick = () => wrapper.find( 'button' ).simulate( 'click' );
          handleClick();

          expect( contextValue.dispatch ).toHaveBeenCalledWith({
               type: types.login,
               payload: {
                    name: 'Pedro Parques'
               }
          });

          expect( mockNavigate ).toHaveBeenCalledWith( '/marvel' , { replace: true } );

          localStorage.setItem( 'lastPath' , '/dc' );

          handleClick();

          expect( mockNavigate ).toHaveBeenCalledWith( '/dc' , { replace: true } )
     })
     
     
})
