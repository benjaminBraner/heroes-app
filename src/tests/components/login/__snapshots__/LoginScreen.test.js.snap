// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el LoginComponent debe mostrarse correctamente 1`] = `
<div
  className="container mt-5"
>
  <h1>
    Login Screen
  </h1>
  <hr />
  <form>
    <input
      autoComplete="off"
      name="inputName"
      onChange={[Function]}
      type="text"
      value=""
    />
    <button
      className="btn btn-primary"
      onClick={[Function]}
      type="submit"
    >
      Login
    </button>
  </form>
</div>
`;
